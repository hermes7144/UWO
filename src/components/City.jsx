import React from 'react';
import { useQuery } from '@tanstack/react-query';
import { getGoods } from '../api/firebase';

export default function City(city) {
  const { data: goods } = useQuery(['goods', city.city], () => getGoods(city.city), { staleTime: Infinity, enabled: !!city.city });
  const { data: nextGoods } = useQuery(['goods', city.nextCity], () => getGoods(city.nextCity), { staleTime: Infinity, enabled: !!city.nextCity });

  const nextItems = nextGoods && nextGoods.map((nextgood) => nextgood.goods_nm);

  return (
    <table>
      <tbody>
        <tr>
          <th>{city.city}</th>

          {goods &&
            goods.map((good, index) => (
              <td key={index} className={nextItems && nextItems.includes(good.goods_nm) ? 'text-red-600' : ''}>
                {good.goods_nm}
              </td>
            ))}
        </tr>
      </tbody>
    </table>
  );
}
