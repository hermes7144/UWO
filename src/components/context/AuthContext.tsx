import { createContext, useContext } from 'react';
import { User } from '@firebase/auth';

export const AuthContext = createContext<User | null>(null);

export function useAuthContext() {
  return useContext(AuthContext);
}
