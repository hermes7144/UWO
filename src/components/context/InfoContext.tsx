import { createContext, useContext } from 'react';

interface IInfoContext {
  getGoods(city: string): Promise<any[]>;
}

export const InfoContext = createContext<IInfoContext | null>(null);

export function useInfoContext() {
  return useContext(InfoContext);
}
