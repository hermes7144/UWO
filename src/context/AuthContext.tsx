import { createContext, useContext } from 'react';
import { User } from '@firebase/auth';

type UserType = {
  user: User & {
    isAdmin?: boolean;
  };
  uid?: string;
  login: () => {};
  logout: () => void;
};

export const AuthContext = createContext<UserType | null>(null);

export function useAuthContext() {
  return useContext(AuthContext);
}
