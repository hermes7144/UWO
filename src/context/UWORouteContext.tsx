import { createContext, useContext } from 'react';
import { Dispatch, SetStateAction } from 'react';

interface IUWORouteContext {
  citys: number[];
  setCitys: Dispatch<SetStateAction<number[]>>;
  coordinates: [number, number] | null;
  setCoordinates: Dispatch<SetStateAction<[number, number] | null>>;
  editable: boolean;
  setEditable: Dispatch<SetStateAction<boolean>>;
}

export const UWORouteContext = createContext<IUWORouteContext | null>(null);

export function useUWORouteContext() {
  return useContext(UWORouteContext);
}
