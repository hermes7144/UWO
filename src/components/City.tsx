import { useEffect } from 'react';
import { BsFillTrashFill } from 'react-icons/bs';
import { useCoordinatesContext } from '../context/CoordinatesContext';
import useCity from '../Hooks/useCity';

type CityProps = {
  citys: number[];
  index: number;
  cityNm: string;
  isEditable?: boolean;
  coordinates: [number, number];
  onDelete: (index: number) => void;
};

export default function City({ citys, index, cityNm, coordinates, isEditable = true, onDelete }: CityProps) {
  const { setCoordinates } = useCoordinatesContext();
  const {
    goodsQuery: { isLoading, data: goods },
  } = useCity(citys[index]);
  const {
    goodsQuery: { data: nextGoods },
  } = useCity(citys[index + 1]);

  const nextItems = nextGoods && nextGoods.map((nextgood) => nextgood.goods_nm);
  const handleDelete = () => onDelete(index);

  const handleClick = (coordinates) => setCoordinates(coordinates);
  useEffect(() => {
    index === 0 && setCoordinates(coordinates);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <section>
      {isLoading && <p>Loading...</p>}
      <table className='border-solid border-2 my-2'>
        <tbody>
          <tr>
            <th className='border-solid border-2 w-6' rowSpan={2}>
              {index + 1}
            </th>
            <th rowSpan={2}>
              <div className='flex flex-col justify-center items-center w-24'>
                <button onClick={() => handleClick(coordinates)}>{cityNm}</button>

                {isEditable && <BsFillTrashFill className='cursor-pointer opacity-50 hover:opacity-100' onClick={handleDelete} />}
              </div>
            </th>
            {goods &&
              goods.map((good, index) => (
                <td key={index} className={'border-solid border-2 w-30 ' + (good.specialty ? 'bg-yellow-200' : '')}>
                  <img className='m-auto' src={good.goods_url} alt='' />
                </td>
              ))}
          </tr>
          <tr>
            {goods &&
              goods.map((good, index) => (
                <td key={index} className='border-solid border-2 text-center text-xs'>
                  <span className={nextItems && nextItems.includes(good.goods_nm) ? 'text-red-600' : ''}>{good.goods_nm}</span>
                </td>
              ))}
          </tr>
        </tbody>
      </table>
    </section>
  );
}
