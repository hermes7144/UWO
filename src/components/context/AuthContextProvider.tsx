import React from 'react';
import { useEffect, useState } from 'react';
import { login, logout, onUserStateChange } from '../../api/firebase';
import { AuthContext } from './AuthContext';
import { User } from '@firebase/auth';

type AuthContextProviderProps = {
  children: React.ReactNode;
};

export function AuthContextProvider({ children }: AuthContextProviderProps) {
  const [user, setUser] = useState<User | null>(null);

  useEffect(() => {
    onUserStateChange((user: User) => {
      setUser(user);
    });
  }, []);

  return <AuthContext.Provider value={{ user, uid: user && user.uid, login, logout }}>{children}</AuthContext.Provider>;
}
