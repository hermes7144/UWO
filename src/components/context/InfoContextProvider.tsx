import { getGoods } from '../../api/firebase';
import { InfoContext } from './InfoContext';
import { useState } from 'react';

type ContextProviderProps = {
  children: React.ReactNode;
};
export function InfoContextProvider({ children }: ContextProviderProps) {
  const [citys, setCitys] = useState<number[]>([]);

  type RouteType = {
    id: string;
    title: string;
    remark: string;
  };
  const [route, setRoute] = useState<RouteType>({ id: '', title: '', remark: '' });

  return <InfoContext.Provider value={{ citys, setCitys, route, setRoute, getGoods }}>{children}</InfoContext.Provider>;
}
