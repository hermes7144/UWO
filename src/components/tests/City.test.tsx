import { render, screen } from '@testing-library/react';
import { Route } from 'react-router-dom';
import { useUWORouteContext } from '../../context/UWORouteContext';
import { fakeGoods } from '../../tests/fakeGoods';
import { withAllContexts, withRouter } from '../../tests/utils';
import City from '../City';

jest.mock('../../context/UWORouteContext');

describe('City', () => {
  const mockUseUWORouteContext = useUWORouteContext as jest.MockedFunction<typeof useUWORouteContext>;
  const fakeUseCity = jest.fn();

  beforeEach(() => {
    mockUseUWORouteContext.mockReturnValue({
      citys: [1],
      coordinates: [1, 3],
      setCitys: jest.fn(),
      setCoordinates: jest.fn(),
      setEditable: jest.fn(),
      editable: true,
    });
  });

  test('renders city name', () => {
    fakeUseCity.mockImplementation(() => {
      return {
        goodsQuery: { isLoading: true, data: fakeGoods },
      };
    });

    renderCity();

    const cityName = screen.getByText(/Seoul/);
    expect(cityName).toBeInTheDocument();
  });

  function renderCity() {
    return render(withAllContexts(withRouter(<Route path='/' element={<City route={{}} index={0} city={{ city_id: 1, city_nm: 'Seoul', city_coordinates: [37.5665, 126.978] }} />} />), fakeUseCity));
  }
});
