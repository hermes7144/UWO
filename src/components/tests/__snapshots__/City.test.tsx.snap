// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`City renders correctly 1`] = `
<DocumentFragment>
  <table
    class="border-solid border-2 mt-2"
  >
    <tbody>
      <tr>
        <th
          class="border-solid border-2 w-6"
          rowspan="2"
        >
          2
        </th>
        <th
          rowspan="2"
        >
          <div
            class="flex justify-center items-center w-24"
          >
            리스보아
            <br />
            <svg
              class="cursor-pointer opacity-50 hover:opacity-100"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M2.5 1a1 1 0 0 0-1 1v1a1 1 0 0 0 1 1H3v9a2 2 0 0 0 2 2h6a2 2 0 0 0 2-2V4h.5a1 1 0 0 0 1-1V2a1 1 0 0 0-1-1H10a1 1 0 0 0-1-1H7a1 1 0 0 0-1 1H2.5zm3 4a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7a.5.5 0 0 1 .5-.5zM8 5a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7A.5.5 0 0 1 8 5zm3 .5v7a.5.5 0 0 1-1 0v-7a.5.5 0 0 1 1 0z"
              />
            </svg>
          </div>
        </th>
        <td
          class="border-solid border-2 w-30"
        >
          <img
            alt=""
            class="m-auto"
            src="http://res.cloudinary.com/da25q6n7o/image/upload/v1678705707/prido8t3x7zzvpmomtoh.png"
          />
        </td>
        <td
          class="border-solid border-2 w-30bg-yellow-200"
        >
          <img
            alt=""
            class="m-auto"
            src="http://res.cloudinary.com/da25q6n7o/image/upload/v1678706021/ln48swpjtphyvymib58z.png"
          />
        </td>
      </tr>
      <tr>
        <td
          class="border-solid border-2 text-center text-xs"
        >
          <span
            class="text-red-600"
          >
            브랜디
          </span>
        </td>
        <td
          class="border-solid border-2 text-center text-xs"
        >
          <span
            class="text-red-600"
          >
            아몬드
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
